### PR TITLE
Support xctrunner devices

### PR DESCRIPTION
**Pull request by bkase**
This PR adds support for XCUITests to be run on devices.

XCUITests use a test-runner app in addition to the app you're trying to
test. Tulsi creates this test runner scheme directly in Xcode from an
ios_ui_tests rule. To run on a device, there are three things needed
that Tulsi didn't support: (1) specifying and injecting the proper
provisioning profile for this test app and (2) using the proper app
identifier in the entitlements that is specified in the provisioning
profile and (3) on Xcode9 there is a new framework that needs to be
re-codesigned along with the other test frameworks.

This diff fixes (1) by supporting a new optional environment variable to
specify the location of a provisioning profile.

(2) is fixed by extracting the app-identifier from the provisioning
profile and swapping some of the $(TeamIdentifier) variables in the
XCTRunner.entitlements template with $(AppIdentifier). As noted in
https://developer.apple.com/library/content/technotes/tn2415/_index.html#//apple_ref/doc/uid/DTS40016427-CH1-ENTITLEMENTSLIST
it is not safe to assume that the team-identifier is the app-identifier.
Since it usually is, we fallback to setting the app-identifier to the
team-identifier if no provisioning profile is provided.

(3) is fixed by adding the new framework to the array of frameworks that
are listed in need of resigning.

### CODE REVIEW

**Comment by shamanskyh** [comment]
LGTM! And really thorough description of the problem and solution, too ✅

File: src/TulsiGenerator/XCTRunner.entitlements [hunk 1]
```diff
@@ -3,14 +3,14 @@
 <plist version="1.0">
 <dict>
 	<key>application-identifier</key>
-	<string>$(TeamIdentifier).$(BundleIdentifier)</string>
+	<string>$(AppIdentifier).$(BundleIdentifier)</string>
```
Comment (jerrymarino) [inline src/TulsiGenerator/XCTRunner.entitlements:6]: Good catch! This seems like a bug? cc @dmishe thoughts?

File: src/TulsiGenerator/Scripts/bazel_build.py [hunk 2]
```diff
@@ -608,6 +613,12 @@ def Run(self, args):
       # the host itself.
       if (self.is_test and self.xcode_version_minor >= 730 and
           self.codesigning_allowed):
+        # If we're codesigning we also need to make sure we use the correct
+        # provisioning profile.
+        if not self.is_simulator and self.test_runner_provisioning_profile:
```
Comment (jerrymarino) [inline src/TulsiGenerator/Scripts/bazel_build.py]: You can remove the check for `self.is_simulator`. The value `self.codesigning_allowed` includes this check baked.

File: src/TulsiGenerator/Scripts/bazel_build.py [hunk 3]
```diff
@@ -1233,10 +1264,53 @@ def _InstantiateUIRunnerEntitlements(self):
...
+      else:
+          contents = contents.replace(
+              '$(AppIdentifier)',
+              self._ExtractSigningTeamIdentifier(self.content_folder_path))
+
       with open(output_file, 'w') as output:
         output.write(contents)
     return output_file
 
+  _app_identifier_for_profile = {}
```
Comment (jerrymarino) [inline src/TulsiGenerator/Scripts/bazel_build.py:1285]: Nitpick: remove the `_` prefix of this variable ( Edit disregard, this is an ivar! )

**Comment by jerrymarino** [comment]
Overall this looks good to me 👍

A few thoughts:
- this is reliant on env vars, which may be tricky to setup. Given the possible requirement, of "building" provisioning profiles it seems fine to put this in the script. Would this be added as another target in the bazel invocation, or another invocation altogether?


- we may want to extract other values from provisioning profile -> entitlements:
    - `aps_environment`
    - `get_task_allow` ( which is typically managed by bazel ). Is there issues doing `Debug/Release` builds?

- it looks like some of the provisioning logic and entitlements derivation may be duplicated here and in `rules_apple` but it seems fine for now since this is the pattern followed

**Pull request merged**